packages/x-data-grid/src/hooks/utils/useGridComponentRenderer.tsx

```diff
@@ -0,0 +1,61 @@
+import * as React from 'react';
+import clsx from 'clsx';
+import { SxProps } from '@mui/system';
+
+export type ComponentRenderFn<Props, State> = (
+  props: Props,
+  state: State,
+) => React.ReactElement<unknown>;
+
+export type RenderProp<State> =
+  | ComponentRenderFn<React.HTMLAttributes<any>, State>
+  | React.ReactElement;
+
+type GridComponentRendererOptions<Props, State> = {
+  defaultElement: keyof JSX.IntrinsicElements | React.ComponentType<Props>;
+  props: Props;
+  state?: State;
+  render?: RenderProp<State>;
+};
+
+/**
+ * Returns a function that renders a component.
+ *
+ * @ignore - internal hook.
+ */
+export function useGridComponentRenderer<
+  Props extends React.HTMLAttributes<any>,
+  State extends Record<string, any>,
+>({
+  defaultElement,
+  props,
+  state = {} as State,
+  render,
+}: GridComponentRendererOptions<Props, State>) {
+  const renderElement = () => {
+    if (typeof render === 'function') {
+      return render(props, state);
+    }
+
+    if (render) {
+      const className = clsx(render.props.className, props.className);
+      const style = { ...render.props.style, ...props.style };
+      const sx = mergeSx(render.props.sx, (props as any).sx);
+      return React.cloneElement(render, { ...props, className, style, sx });
+    }
+
+    return React.createElement(defaultElement, props);
+  };
+
+  return { renderElement };
```

Comment:
The object + function wrapper could be removed to just return the JSX, it would create less work for the garbage collector.